File: Java/civitas/src/civitas/Jugador.java
package civitas;

import GUI.Dado;
import java.util.ArrayList;
import java.lang.Comparable;

public class Jugador implements Comparable<Jugador>{
    private String nombre;
    protected boolean encarcelado=false;
    protected static final int casasMax = 4;
    protected static final int casasPorHotel = 4;
    protected static final int hotelesMax = 4;
    private int numCasillaActual = 0;
    protected static final float pasoPorSalida = (float) 1000.0;
    protected static final float precioLibertad = (float) 200.0;
    private boolean puedeComprar;
    private float saldo = saldoInicial;
    private static final float saldoInicial = (float) 7500.0;
    private ArrayList<TituloPropiedad> propiedades;
    private SorpresaSalirCarcel salvoConducto = null;
    
    
    public Jugador(String _nombre){
        this.nombre=_nombre;
        propiedades = new ArrayList<>();
        saldo = saldoInicial;
        numCasillaActual=0;
        puedeComprar = false;
    } 
    
    Jugador(Jugador _jugador){
        this.nombre=_jugador.getNombre();
        this.encarcelado=_jugador.isEncarcelado();
        this.numCasillaActual=_jugador.getNumCasillaActual();
        this.puedeComprar=_jugador.getPuedeComprar();
        this.saldo=_jugador.getSaldo();
        this.propiedades=_jugador.getPropiedades();
        this.salvoConducto=_jugador.salvoConducto;
    }

    public String getNombre() {
        return nombre;
    }

    public boolean isEncarcelado() {
        return encarcelado;
    }

    public int getCasasMax() {
        return casasMax;
    }

    public int getCasasPorHotel() {
        return casasPorHotel;
    }

    public int getHotelesMax() {
        return hotelesMax;
    }

    public int getNumCasillaActual() {
        return numCasillaActual;
    }

    public static float getPasoPorSalida() {
        return pasoPorSalida;
    }

    public static float getPrecioLibertad() {
        return precioLibertad;
    }

    public boolean getPuedeComprar() {
        return puedeComprar;
    }

    public float getSaldo() {
        return saldo;
    }

    public static float getSaldoInicial() {
        return saldoInicial;
    }

    public ArrayList<TituloPropiedad> getPropiedades() {
        return propiedades;
    }
    
    public boolean tieneAlgoQueGestionar(){
        if(propiedades.isEmpty())
            return false;
        else
            return true;
    }
    
    
    public boolean paga(float cantidad){
        if(cantidad<0)
            return this.modificarSaldo(cantidad);
        else
            return this.modificarSaldo(cantidad*(-1));
    }
    public boolean modificarSaldo(float cantidad){
        this.saldo+=cantidad;
        Diario.getInstance().ocurreEvento("Se modifica el saldo");
        return true;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }
    
    

    protected boolean cancelarHipoteca(int _ip){
        boolean result = false;
        if(this.encarcelado){
            return result;
        }
        if(this.existeLaPropiedad(_ip)){
            TituloPropiedad propiedad=propiedades.get(_ip);
            float cantidad = propiedad.getImporteCancelarHipoteca();
            boolean _puedoGastar=this.puedoGastar(cantidad);
            if(_puedoGastar){
                result = propiedad.cancelarHipoteca(this);
                if(result){
                    Diario.getInstance().ocurreEvento("El jugador " + this.nombre + " cancela la hipoteca de la propiedad " + _ip);
                }
            }
        }
        return result;
    }

    public void setPuedeComprar(boolean puedeComprar) {
        this.puedeComprar = puedeComprar;
    }
    
    
    
    protected int cantidadCasasHoteles(){
        int nC=0, nH=0, total;
        for(int i=0; i<this.propiedades.size(); i++){
            nC+=this.propiedades.get(i).getNumCasas();
            nH+=this.propiedades.get(i).getNumHoteles();
        }
        total=nC+nH;
        return total;
    }
    @Override
    public int compareTo(Jugador otroJugador){
        int saldoOtro=(int) ((Jugador) otroJugador).getSaldo();
        return (int) (saldoOtro-this.getSaldo());    
    }
    

    protected boolean comprar(TituloPropiedad titulo){
        boolean result = false;
        if(encarcelado==false && titulo.tienePropietario()==false){
            if(puedeComprar){
                float precio = titulo.getPrecioCompra();
                if(puedoGastar(precio)){
                    result = titulo.comprar(this);
                    if(result){
                        propiedades.add(titulo);
                        Diario.getInstance().ocurreEvento("El jugador " + nombre + " ha comprado " + titulo.toString());
                    }
                    puedeComprar=false;
                }
            }
        }
        return result;
    }

    protected boolean construirHotel(int _ip){
        boolean result = false;
        if(this.encarcelado){
            return result;
        }
        if(this.existeLaPropiedad(_ip)){
            TituloPropiedad propiedad = propiedades.get(_ip);
            boolean puedoEdificarHotel = this.puedoEdificarHotel(propiedad);
            float precio = propiedad.getPrecioEdificar();
            
            if(this.puedoGastar(precio) && propiedad.getNumHoteles()<this.getHotelesMax() && propiedad.getNumCasas()>=this.getCasasPorHotel()){
                puedoEdificarHotel=true;
            }
            if(puedoEdificarHotel){
                result = propiedad.construirHotel(this);
                int casasPorHotel = this.getCasasPorHotel();
                propiedad.derruirCasa(casasPorHotel, this);
                if(result){
                    Diario.getInstance().ocurreEvento("El jugador " + this.nombre + " construye hotel en la propiedad " + _ip);
                }
            }
        }
        return result;
    }

    protected boolean construirCasa(int _ip){
        boolean result = false;
        if(this.encarcelado){
            return result;
        }
        if(this.existeLaPropiedad(_ip)){
            TituloPropiedad propiedad = propiedades.get(_ip);
            boolean puedoEdificarCasa = this.puedoEdificarCasa(propiedad);
            float precio = propiedad.getPrecioEdificar();
            
            if(this.puedoGastar(precio) && propiedad.getNumCasas()<this.getCasasMax()){
                puedoEdificarCasa=true;
            }
            if(puedoEdificarCasa){
                result = propiedad.construirCasa(this);
                if(result){
                    Diario.getInstance().ocurreEvento("El jugador " + this.nombre + " construye casa en la propiedad " + _ip);
                }
            }
        }
        return result;
    }
    
    protected boolean debeSerEncarcelado(){
        if(this.isEncarcelado())
            return false;
        else{
            if(!this.tieneSalvoconducto())
                return true;
            else{
                this.perderSalvoconducto();
                Diario.getInstance().ocurreEvento("Jugador salvado de la carcel");
                return false;
            }
            
        }
    }
    
    boolean enBancarrota(){
        if(this.saldo<0)
            return true;
        else
            return false;
    }
    
    boolean encarcelar(int _numCasillaCarcel){
        if(this.debeSerEncarcelado()){
            this.moverACasilla(_numCasillaCarcel);
            this.encarcelado=true;
            Diario.getInstance().ocurreEvento("Jugador encarcelado");
        }
        return this.encarcelado;
    }
    
    private boolean existeLaPropiedad(int _ip){
        boolean solucion= false;
        for(int i=0; i<propiedades.size(); i++){
		if(_ip==i)
			solucion =  true;
		
                
	}
        return solucion;
        
    }

    boolean hipotecar(int _ip){
        boolean result = false;
        if(this.encarcelado){
            return result;
        }
        if(this.existeLaPropiedad(_ip)){
            TituloPropiedad propiedad = propiedades.get(_ip);
            result = propiedad.hipotecar(this);
        }
        if(result){
            Diario.getInstance().ocurreEvento("El jugador " + this.nombre + " hipoteca la propiedad " + _ip);
        }
        return result;
    }
    
    boolean moverACasilla(int _numCasilla){
        if(this.isEncarcelado())
            return false;
        else{
            this.numCasillaActual=_numCasilla;
            this.puedeComprar=false;
            Diario.getInstance().ocurreEvento("Se ha movido a la casilla " + _numCasilla);
            return true;
        }
    }
    
    boolean obtenerSalvoconducto(SorpresaSalirCarcel _sorpresa){
        if(this.isEncarcelado())
            return false;
        else{
            this.salvoConducto=_sorpresa;
            return true;
        }
    }
    
    boolean pagaAlquiler(float _cantidad){
        if(this.encarcelado)
            return false;
        else{
            return this.paga(_cantidad);
        }
    }
    
    boolean pagaImpuesto(float _cantidad){
        if(this.encarcelado)
            return false;
        else{
            return this.paga(_cantidad);
        }
    }
    
    boolean pasaPorSalida(){
        this.modificarSaldo(pasoPorSalida);
        Diario.getInstance().ocurreEvento("Se pasa por la casilla de salida");
        return true;
    }
    
    void perderSalvoconducto(){
        salvoConducto.usada();
        this.salvoConducto=null;
    }
    
    boolean puedeComprarCasilla(){
        if(this.encarcelado==true)
            this.puedeComprar=false;
        else
            this.puedeComprar=true;
        
        return this.puedeComprar;
    }
    
    private boolean puedeSalirCarcelPagando(){
        if(this.saldo>=precioLibertad)
            return true;
        else
            return false;
    }
    
    private boolean puedoEdificarCasa(TituloPropiedad _propiedad){
        if(saldo>=_propiedad.getPrecioEdificar() && this.nombre.equals(_propiedad.getPropietario().getNombre()))
            if(_propiedad.getNumCasas()<4)
                return true;
            else
                return false;
        else
            return false;
        
    }
    
    private boolean puedoEdificarHotel(TituloPropiedad _propiedad){
        if(saldo>=_propiedad.getPrecioEdificar() && this.nombre.equals(_propiedad.getPropietario().getNombre()))
            if(_propiedad.getNumCasas()==4 && _propiedad.getNumHoteles()<4)
                return true;
            else
                return false;
        else
            return false;
    }
    
    private boolean puedoGastar(float precio){
        if(this.isEncarcelado())
            return false;
        else{
            if(this.saldo>=precio)
                return true;
            else
                return false;
        }
    }
    
    boolean recibe(float cantidad){
         if(this.encarcelado==true)
             return false;
         else{
             this.modificarSaldo(cantidad);
             return true;
         }
     }
     
    boolean salirCarcelPagando(){
        if(this.isEncarcelado() && this.getSaldo()>=precioLibertad){
            this.paga(precioLibertad);
            this.encarcelado=false;
            Diario.getInstance().ocurreEvento("El jugador sale de la carcel pagando");
            return true;
        }
        else
            return false;
    } 
    
    boolean salirCarcelTirando(){
        boolean resultado = Dado.getInstance().salgoDeLaCarcel();
        if(resultado==true){
            this.encarcelado=false;
            Diario.getInstance().ocurreEvento("Sale de la carcel tirando");
        }
        return resultado;
            
    }
    
    boolean tieneSalvoconducto(){
        if(this.salvoConducto==null)
            return false;
        else
            return true;
    }
    
     boolean vender(int _ip){
        if(this.encarcelado==true)
             return false;
         else{
             this.existeLaPropiedad(_ip);
             if(this.propiedades.get(_ip).vender(this)){
                 this.propiedades.remove(_ip);
                 Diario.getInstance().ocurreEvento("Se ha vendido una propiedad");
                 return true;
             }
             else
                return false;
         }
    }
    
    @Override
    public String toString(){
        String texto="";
        texto= "Jugador { Nombre: "+ nombre + " ,saldo: " + saldo;
        if(encarcelado)
            texto+=", estoy encarcelado.";
        else
            texto+=", no estoy encarcelado.";
        
        if(propiedades.isEmpty()){
            texto += " No tienes propiedades... (Por ahora) ";
        }
        else{
            texto+=" Mis propiedades son: \n";
        
            for(TituloPropiedad p: propiedades){
                texto+=p.toString() + "\n";
            }
        }
        
        return texto;
    }
    
    
    public boolean soyEspeculador(){
        return false;
    }
   
       
    
    //EXAMEN
   public void robarPropiedad(TituloPropiedad _titulo){
       _titulo.actualizaPropietarioPorConversion(this);
   }
}

File: Ruby/civitas/pkg/civitas-0.0.1/lib/jugador.rb
#encoding: utf­8

require_relative 'dado'
require_relative 'diario'
require_relative 'titulo_propiedad'
require_relative 'sorpresa'
module Civitas
  class Jugador
    
    attr_accessor :casas_max, :casas_por_hotel, :salvo_conducto, :hoteles_max, :encarcelado, :paso_por_salida, :precio_libertad, :nombre, :puede_comprar, :saldo, :num_casilla_actual, :propiedades
    
    @@casas_max=4
    @@casas_por_hotel=4
    @@hoteles_max=4
    @@paso_por_salida=1000
    @@precio_libertad=20000
    
    @@saldo_inicial=7500
    
    def initialize(_nombre)
      @encarcelado=false
      @nombre=_nombre
      @num_casilla_actual=0
      @puede_comprar=false
      @saldo=7500.0
      @propiedades= Array.new()
      @salvo_conducto=nil
    end
    
    def self.new_jugador_copia(_jugador)
      jugador = new(_jugador.nombre)
      jugador.saldo=_jugador.saldo
      jugador.encarcelado=_jugador.encarcelado
      jugador.num_casilla_actual=_jugador.num_casilla_actual
      jugador.puede_comprar= _jugador.puede_comprar
      jugador.propiedades= _jugador.propiedades
      jugador.salvo_conducto=_jugador.salvo_conducto
      return jugador
    end
    

    def cancelar_hipoteca(_ip)
      result = false
      if(@encarcelado==true)
        return result
      end
      if(self.existe_la_propiedad(_ip))
        propiedad = @propiedades.at(_ip)
        cantidad = propiedad.get_importe_cancelar_hipoteca()
        _puedo_gastar = self.puedo_gastar(cantidad)
        if(_puedo_gastar)
          result = propiedad.cancelar_hipoteca(self)
          if(result)
            Diario.instance.ocurre_evento("El jugador " + self.nombre + " cancela la hipoteca de la propiedad " + _ip.to_s)
          end
        end
      end
      return result
    end
    
    def cantidad_casas_hoteles()
      nc=0
      nh=0
      for i in 0..(@propiedades.size()-1)
        nc+=@propiedades.at(i).get_num_casas()
        nh+=@propiedades.at(i).get_num_hoteles()
      end
      total=nc+nh
      return total
    end
    
    def comprar(_titulo)
      result = false
      if(@encarcelado==true)
        return result
      end
      if(@puede_comprar==true)
        precio = _titulo.precio_compra
        if(puedo_gastar(precio))
          result = _titulo.comprarT(self)
        end
        if(result)
          @propiedades << _titulo
          Diario.instance.ocurre_evento("El jugador " + self.nombre + " compra la propiedad " + _titulo.to_st());
        end
        @puede_comprar=false
      end
      return result
    end
    
    def construir_casa(_ip)
      result = false
        if(@encarcelado)
            return result
        end
        if(self.existe_la_propiedad(_ip))
            propiedad = @propiedades.at(_ip)
            puedo_edificar_casa = self.puedo_edificar_casa(propiedad)
            precio = propiedad.precio_edificar
            
            if(self.puedo_gastar(precio) && propiedad.num_casas<@@casas_max)
                puedo_edificar_casa = true
            end
            if(puedo_edificar_casa)
                result = propiedad.construir_casa(self)
                if(result)
                    Diario.instance.ocurre_evento("El jugador " + self.nombre + " construye casa en la propiedad " + _ip.to_s());
                end
                end
        end
        return result;
    end
    
    def construir_hotel(_ip)
      result = false
        if(@encarcelado)
            return result
        end
        if(self.existe_la_propiedad(_ip))
            propiedad = @propiedades.at(_ip)
            puedo_edificar_hotel = self.puedo_edificar_hotel(propiedad)
            precio = propiedad.precio_edificar
            
            if(self.puedo_gastar(precio) && propiedad.num_hoteles<@@hoteles_max && propiedad.num_casas>=@@casas_por_hotel)
                puedo_edificar_hotel =true
            end
            if(puedo_edificar_hotel)
                result = propiedad.construir_hotel(self)
                _casas_por_hotel = self.casas_por_hotel
                propiedad.derruir_casa(_casas_por_hotel, self)
                if(result)
                    Diario.instance.ocurre_evento("El jugador " + self.nombre + " construye hotel en la propiedad " + _ip.to_s);
                end
                end
        end
        return result;
    end
    
    def debe_ser_encarcelado()
      if(self.encarcelado)
        return false
      else
        if(!self.tiene_salvoconducto())
          return true
        else
          self.perder_salvoconducto()
          Diario.instance.ocurre_evento("Jugador salvado de la carcel")
          return false
        end
      end
    end
    
    def es_bancarrota()
      if(@saldo < 0)
        return true
      else
        return false
      end
    end
    
    def encarcelar(_num_casilla_carcel)
      if(self.debe_ser_encarcelado())
        self.mover_a_casilla(_num_casilla_carcel)
        self.encarcelado=true
        Diario.instance.ocurre_evento("Jugador encarcelado")
      end
      return @encarcelado
    end
    
    def existe_la_propiedad(_ip)
      solucion=false
      for i in 0..(@propiedades.size()-1)
        if(i==_ip)
          solucion = true
        end
      end
      return solucion
    end
    
    def hipotecar(_ip)
      result = false
        if(self.encarcelado)
            return result
        end
        if(self.existe_la_propiedad(_ip))
            propiedad = @propiedades.at(_ip)
            result = propiedad.hipotecarT(self)
        end
        if(result)
            Diario.instance.ocurre_evento("El jugador " + self.nombre + " hipoteca la propiedad " + _ip.to_s())
        end
        return result
    end
    
    
    def modificar_saldo(_cantidad)
      puts "En modifica Saldo " + @saldo.inspect
      @saldo =  @saldo + _cantidad.to_f()
      Diario.instance.ocurre_evento("Se modifica el saldo")
      return true
    end
    
    def mover_a_casilla(_num_casilla)
      if(@encarcelado==true)
        return false
      else
        @num_casilla_actual=_num_casilla
        @puede_comprar=false
        Diario.instance.ocurre_evento("Se ha movido a la casilla" + @num_casilla_actual.to_s())
        return true
      end
    end
    
    def obtener_salvoconducto(_sorpresa)
      if(self.encarcelado)
        return false
      else
        @salvoconducto=_sorpresa
        return true
      end
    end
    
    def paga(_cantidad)
      return modificar_saldo(-_cantidad)
    end
    
    def paga_alquiler(_cantidad)
      if(@encarcelado)
        return false
      else
        return self.paga(_cantidad)
      end
    end
    
    def paga_impuesto(_cantidad)
      if(@encarcelado)
        return false
      else
        return self.paga(_cantidad)
      end
    end
    
    def pasa_por_salida()
      self.modificar_saldo(@@paso_por_salida)
      Diario.instance.ocurre_evento("Pasaste por la casilla de Salida")
      return true
    end
    
    def perder_salvoconducto()
      @salvoconducto.usada()
      @salvoconducto=nil
    end
    
    def puede_comprar_casilla()
      if(@encarcelado)
        @puede_comprar=false
      else
        @puede_comprar=true
      end
      return @puede_comprar
    end
    
    def puede_salir_carcel_pagando()
      if(@saldo>=@@precio_libertad)
        return true
      else
        return true
      end
    end
    
    def puedo_edificar_casa(_propiedad)
      if(@saldo>=_propiedad.precio_edificar && @nombre.eql?(_propiedad.propietario.nombre))
        if(_propiedad.num_casas<4)
          return true
        else
          return false
        end
      else
        return false
      end
    end
    
    def puedo_edificar_hotel(_propiedad)
      if(@saldo>=_propiedad.precio_edificar && @nombre.eql?(_propiedad.propietario.nombre))
        if(_propiedad.num_casas==4 && _propiedad.num_hoteles<4)
          return true
        else
          return false
        end
      else
        return false
      end
    end
    
    def puedo_gastar(_precio)
      if(@encarcelado)
        return false
      else
        if(@saldo>=_precio)
          return true
        else
          return false
        end
      end
    end
    
    def recibe(_cantidad)
      if(@encarcelado==true)
        return false
      else
        self.modificar_saldo(_cantidad)
        return true;
      end
    end
    
    def salir_carcel_pagando()
      if(@encarcelado && @saldo>=@@precio_libertad)
        self.paga(@@precio_libertad)
        @encarcelado=false
        Diario.instance.ocurre_evento("Sale de la carcel pagando")
        return true
      else
        return false
      end
    end
    
    def salir_carcel_tirando()
      puedo_salir=Dado.new.salgo_de_la_carcel()
      if(puedo_salir==true)
        @encarcelado=false
        Diario.instance.ocurre_evento("Sale de la carcel tirando")
      end
      return puedo_salir
      
    end
    
    def tiene_algo_que_gestionar()
      if(@propiedades.length() == 0)
          return false
      else
        return true
      end
    end
    
    def tiene_salvoconducto()
      if(@salvoconducto==nil)
        return false
      else
        return true
      end
    end
    
    def to_sj()
      texto = "Jugador { Nombre: #{@nombre}, saldo: #{@saldo} "
      if(self.encarcelado)
        texto+=" y estoy encarcelado"
      else
        texto+=" y no estoy encarcelado"
      end
      texto+=" Mis propiedades son: "
      if(@propiedades1!=nil)
        texto+=" y no tengo propiedades"
      else
        for i in @propiedades
        texto+=i.to_st()
        end
      end
      return texto
    end
    
    def vender(_ip)
      if(@encarcelado==true)
        return false
      else
        if(self.existe_la_propiedad(_ip))
          if(@propiedades.at(_ip).hipotecado==false)
            resul = @propiedades.at(_ip).venderT(self)
            @propiedades.delete_at(_ip)
            Diario.instance.ocurre_evento("Se vendio una propiedad")
            return resul
          else
            return false          
          end
        end
        return false
      end
    end
    
    def set_puede_comprar(_puedecomprar)
      @puede_comprar = _puedecomprar
    end
    
    def compare_to(otro)
      return @saldo <=> otro.saldo
    end
    
    
  end
end

